core/src/main/java/org/ehcache/UserManagedCacheBuilder.java

```diff
@@ -94,23 +106,94 @@ T build(ServiceLocator serviceLocator) throws IllegalStateException {
     }
     final Store.Provider storeProvider = serviceLocator.getService(Store.Provider.class);
 
-    Store.Configuration<K, V> storeConfig = new StoreConfigurationImpl<K, V>(keyType, valueType,
-        evictionVeto, evictionPrioritizer, classLoader, expiry, resourcePools);
+    List<LifeCycled> lifeCycledList = new ArrayList<LifeCycled>();
+    
+    Store.Configuration<K, V> storeConfig;
     boolean persistent = resourcePools.getResourceTypeSet().contains(ResourceType.Core.DISK);
+    final Store<K, V> store;
     if (persistent) {
       if (id == null) {
         throw new IllegalStateException("Persistent user managed caches must have an id set");
       }
-      storeConfig = new PersistentStoreConfigurationImpl<K, V>(storeConfig, id);
+      LocalPersistenceService persistenceService = serviceLocator.getService(LocalPersistenceService.class);
+      if (!resourcePools.getPoolForResource(ResourceType.Core.DISK).isPersistent()) {
+        try {
+          persistenceService.destroyPersistenceSpace(id);
+        } catch (CachePersistenceException cpex) {
+          throw new RuntimeException("Unable to clean-up persistence space for non-restartable cache " + id, cpex);
+        }
+      }
+      try {
+        LocalPersistenceService.PersistenceSpaceIdentifier space = persistenceService.getOrCreatePersistenceSpace(id);
+        SerializationProvider serialization = serviceLocator.getService(SerializationProvider.class);
+        Serializer<K> keySerializer;
+        Serializer<V> valueSerializer;
+        try {
+          keySerializer = serialization.createKeySerializer(keyType, classLoader, space);
+          try {
+            valueSerializer = serialization.createValueSerializer(valueType, classLoader, space);
+          } catch (UnsupportedTypeException e) {
+            try {
+              keySerializer.close();
+            } catch (IOException f) {
+              LOGGER.warn("Exception shutting down key serializer, after value serializer was unavailable.", f);
+            } finally {
+              keySerializer = null;
+            }
+            throw e;
+          }
+          lifeCycledList.add(closerFor(keySerializer));
+          lifeCycledList.add(closerFor(valueSerializer));
+        } catch (UnsupportedTypeException e) {
+          throw new RuntimeException(e);
+        }
+        storeConfig = new StoreConfigurationImpl<K, V>(keyType, valueType, evictionVeto, evictionPrioritizer, classLoader,
+                expiry, resourcePools, keySerializer, valueSerializer);
+        store = storeProvider.createStore(storeConfig, space);
+      } catch (CachePersistenceException cpex) {
+        throw new RuntimeException("Unable to create persistence space for cache " + id, cpex);
+      }
+    } else {
+      Serializer<K> keySerializer = null;
```

Comment:
looks a lot like the previous block. Can't that be extracted out of the if / else?